### ISSUE TITLE
struct containing `u128` fields across architecture does not align correctly.

### ISSUE THREAD

**Issue by Myriad-Dreamin** [rkyv/rkyv#409]
We are using rkyv as serialization library for ipc between host and wasm plugin. The bad alignment happened when arm host is sending data to the wasm plugin. Specifically, the following structure caused the issue:

```rs
/// Data of an `<image/>` element.
#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "rkyv", derive(Archive, rDeser, rSer))]
#[cfg_attr(feature = "rkyv-validation", archive(check_bytes))]
pub struct Image {
    /// The encoded image data.
    pub data: Vec<u8>,
    /// The format of the encoded `buffer`.
    pub format: ImmutStr,
    /// The size of the image.
    pub size: Axes<u32>,
    /// A text describing the image.
    pub alt: Option<ImmutStr>,
    /// prehashed image content.
    pub hash: u128,
}
```

In our scenario, `align_of::<Image>` is 16 in arm host, however is 8 in the wasm plugin.

To fix our issue, we changed the `hash` field to type `(u64, u64)`. But this hints that `rkyv` cannot work well in default setting across all architecture. This introduces a problem that how to use `rkyv` practically.

Related issue: [Failed to preview with figures on arm64 device](https://github.com/Enter-tainer/typst-preview/issues/77)

**Comment by djkoloski** [comment]
Sorry about this, this has been a longstanding issue in 0.7.* that we haven't been able to fix due to backwards compatibility requirements. I plan to fix this for 0.8.0 at the cost of some ergonomics.

**Comment by Myriad-Dreamin** [comment]
> Sorry about this, this has been a longstanding issue in 0.7.* that we haven't been able to fix due to backwards compatibility requirements. I plan to fix this for 0.8.0 at the cost of some ergonomics.

Thank you for quick reply. I'm happy to hear that it would be fixed in 0.8.0.

And back to `u128` problem, is there helper (or tutorial) to help detect and avoid these problems?
+ For helper example, I could configure something so that rust compiler emits an hard error when I'm trying to add a `u128` field to that struct.
+ For tutorial example, Some documentation teachs me a list of thing that I cannot do in `rkyv`.

I feel lucky that I was saved by `rkyv`'s cool validation feature, but I should also learn to bypass these limitations.

**Comment by djkoloski** [comment]
I think one good option could be to emit the size and alignment of all of your primitives. Then just assert that they match what you expect in your program. For example, on an x64 machine:

```rust
macro_rules! layout {
    ($($prim:ty),* $(,)?) => {
        $(layout!(@one $prim);)*
    };
    (@one $prim:ty) => {{
        println!("assert_eq!(core::mem::size_of::<{}>(), {});", stringify!($prim), core::mem::size_of::<$prim>());
        println!("assert_eq!(core::mem::align_of::<{}>(), {});", stringify!($prim), core::mem::align_of::<$prim>());
    }};
}

fn main() {
    layout!(
        (),
        bool,
        u8, u16, u32, u64, u128,
        i8, i16, i32, i64, i128,
        f32, f64,
        char,
    );
}
```

Generates:

```
assert_eq!(core::mem::size_of::<()>(), 0);
assert_eq!(core::mem::align_of::<()>(), 1);
assert_eq!(core::mem::size_of::<bool>(), 1);
assert_eq!(core::mem::align_of::<bool>(), 1);
assert_eq!(core::mem::size_of::<u8>(), 1);
assert_eq!(core::mem::align_of::<u8>(), 1);
assert_eq!(core::mem::size_of::<u16>(), 2);
assert_eq!(core::mem::align_of::<u16>(), 2);
assert_eq!(core::mem::size_of::<u32>(), 4);
assert_eq!(core::mem::align_of::<u32>(), 4);
assert_eq!(core::mem::size_of::<u64>(), 8);
assert_eq!(core::mem::align_of::<u64>(), 8);
assert_eq!(core::mem::size_of::<u128>(), 16);
assert_eq!(core::mem::align_of::<u128>(), 8);
assert_eq!(core::mem::size_of::<i8>(), 1);
assert_eq!(core::mem::align_of::<i8>(), 1);
assert_eq!(core::mem::size_of::<i16>(), 2);
assert_eq!(core::mem::align_of::<i16>(), 2);
assert_eq!(core::mem::size_of::<i32>(), 4);
assert_eq!(core::mem::align_of::<i32>(), 4);
assert_eq!(core::mem::size_of::<i64>(), 8);
assert_eq!(core::mem::align_of::<i64>(), 8);
assert_eq!(core::mem::size_of::<i128>(), 16);
assert_eq!(core::mem::align_of::<i128>(), 8);
assert_eq!(core::mem::size_of::<f32>(), 4);
assert_eq!(core::mem::align_of::<f32>(), 4);
assert_eq!(core::mem::size_of::<f64>(), 8);
assert_eq!(core::mem::align_of::<f64>(), 8);
assert_eq!(core::mem::size_of::<char>(), 4);
assert_eq!(core::mem::align_of::<char>(), 4);
```

**Comment by djkoloski** [comment]
This is fixed in 0.8 since bare primitive types are no longer used.
